scrapy_autoextract/providers.py

```diff
@@ -1,114 +1,216 @@
 
     @classmethod
-    def get_page_type(cls) -> str:
-        """Page type is defined by the class attribute `item_key` available on
-        `autoextract_poet.page_inputs` classes.
-        """
-        return cls.provided_class.item_key
+    def provided_classes(cls, type_: Callable) -> bool:
+        return (inspect.isclass(type_)
+                and issubclass(type_, (AutoExtractData, AutoExtractHtml)))
 
+    def __init__(self, crawler: Crawler):
+        """Initialize provider storing its dependencies as attributes."""
+        self.crawler = crawler
+        self.settings = crawler.spider.settings
+        self.logger = self.crawler.spider.logger
+        self.task_manager = get_autoextract_task_manager(crawler)
+        self.aiohttp_session = self.create_aiohttp_session()
+        retries_count = self.settings.getint(
+                "AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES", 3)
+        self.common_request_kwargs = dict(
+            api_key=self.settings.get("AUTOEXTRACT_USER"),
+            endpoint=self.settings.get("AUTOEXTRACT_URL"),
+            max_query_error_retries=retries_count,
+            session=self.aiohttp_session
+        )
+        per_domain_concurrency = get_concurrent_requests_per_domain(self.settings)
+        self.per_domain_semaphore = SlotsSemaphore(per_domain_concurrency)
+        self.logger.info(
+            f"AutoExtractProvider started. Retries: {retries_count}, "
+            f"per domain concurrency: {per_domain_concurrency}"
+        )
 
-class ArticleDataProvider(_Provider):
-
-    provided_class = AutoExtractArticleData
+    def create_aiohttp_session(self) -> aiohttp.ClientSession:
+        concurrent_connections = self.settings.get("CONCURRENT_REQUESTS", 16)
+        self.logger.info(
+            f"AutoExtractProvider concurrent requests: {concurrent_connections}"
+        )
+        return create_session(connection_pool_size=concurrent_connections)
 
+    def create_retry_wrapper(self):
+        return RetryFactory().build()
 
-class ProductDataProvider(_Provider):
+    async def do_request(self, *args, **kwargs):
+        return await request_raw(*args, **kwargs)
 
-    provided_class = AutoExtractProductData
+    def pre_process_item_data(self, data) -> Any:
+        """
+        Hook for transforming data before any further processing, just after
+        receive the content
+        """
+        return data
 
+    def get_per_domain_concurrency_slot(self, request: ScrapyRequest) -> Hashable:
+        """
+        Return the key that will be used to identify the domain of this request.
+        This key is used to modulate the per request concurrency that can be
+        set using the setting `CONCURRENT_REQUESTS_PER_DOMAIN`.
 
-def install():
-    """Register all providers for their respective provided classes."""
-    ArticleDataProvider.register()
-    ProductDataProvider.register()
+        By default the key is the request domain. Override it to change
+        the behavior.
```

Comment:
I wonder if it should use `request.meta['download_slot']` if it is set

Reply:
Good point. Let's think about it. This is the Scrapy code that computes the final `download_slot` key:

https://github.com/scrapy/scrapy/blob/0e579182319504ba7bfd0c09333fe92f70c6d312/scrapy/core/downloader/__init__.py#L108-L116

Ideally, we should just mimic the behaviour in the provider, even calling the same function to compute the key. The later doesn't look like the right thing: it would involve accessing the downloader from a provider and invoke a private function. 

We might copy the code directly here in the function as:
```py
def get_per_domain_concurrency_slot(self, request: ScrapyRequest) -> Hashable:
        if 'download_slot' in request.meta:
            return request.meta['download_slot']

        key = urlparse_cached(request).hostname or ''
        if self.ip_concurrency:
            key = dnscache.get(key, key)
        return key
```

`self.ip_concurrency` would be `self.settings.getint('CONCURRENT_REQUESTS_PER_IP')`

If at some point, you need to decouple from this policy, you can just override the provider and implement this method differently. This way `tldextract` wouldn't be a requirement for this project anymore. 

It seems to work. What do you think? I'm going to implement it.

Reply:
@kmike I have implemented the logic in a different PR (https://github.com/scrapinghub/scrapy-autoextract/pull/17), because now I'm unsure. Particularly with the per IP logic. Look to this line: https://github.com/scrapinghub/scrapy-autoextract/pull/17/files#diff-64092c6e77be2d563b1b97477967d8306ea8901a9aa46db4023465ecd64d4559R147

This is accessing just a DNS cache... But logic to resolve DNS and populate the cache runs separately (https://docs.scrapy.org/en/latest/topics/settings.html?highlight=dns#std-setting-DNS_RESOLVER). I'm unsure about if we would get AutoExtract URLs resolved and populated automatically there. Do you know it? I would say it is not the case, as we are not using Twisted there for anything. 

So it seems that implementing per IP policy wouldn't be as easy.

Reply:
@kmike I'm thinking again on all this. It makes no sense to implement IP policy or `download_slot` here at all. The idea here is to be aligned with what AutoExtract API is using internally for domain occupy, and this is the domain as is defined in this PR. I don't see why it could be useful to implement a different policy here. So +1 to leave the PR as is.

Reply:
If the behavior is so different, I wonder if it'd be better not to reuse CONCURRENT_REQUESTS_PER_DOMAIN option, but add a different option, AUTOEXTRACT_CONCURRENT_REQUESTS_DOMAIN.

Reply:
Can be done later though.

Reply:
Implement ``AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN`` setting to control the concurrency of AutoExtract requests separately for a more flexible configuration. If not set, the provider will search for the `CONCURRENT_REQUESTS_PER_DOMAIN` setting instead. Defaults to ``8``.

Reply:
Thank you @sortafreel for this improvement. However, the new behaviour is that if both `AUTOEXTRACT_CONCURRENT_REQUESTS_PER_DOMAIN` and `CONCURRENT_REQUESTS_PER_DOMAIN` are missing, an error is raised. But it should just default to 8 instead. I see that this was lost in this commit: https://github.com/scrapinghub/scrapy-autoextract/pull/16/commits/692fbe430ea85c478ba50fd9f35be1956610fca8

Reply:
If I'm not mistaken, `CONCURRENT_REQUESTS_PER_DOMAIN` can't be missing, as it is a built-in Scrapy setting; default is set here: https://github.com/scrapy/scrapy/blob/0a1e2fefab04984a0f3c2b470346fd9f2ffde65b/scrapy/settings/default_settings.py#L44. So I adviced @sortafreel to remove the default 8, to avoid duplication.

Reply:
Ouch. I see, thank you @kmike . So I was wrong. I think I saw some code in the Scrapy like `self.settings.getint('CONCURRENT_REQUESTS_PER_DOMAIN', 8)` but I cannot find it now so probably I'm confusing things.